File: Matrices_Python/src/Matrix.py
from __future__ import annotations
from typing import List, Union
from multipledispatch import dispatch
import copy


class Matrix:
    """ Implementation of matrices """

    __matrix: List[List[float]]  # matrix
    __rows: int  # rows
    __columns: int  # columns

    @dispatch()
    def __init__(self) -> None:
        self.__matrix = [[0]]
        self.__rows = 1
        self.__columns = 1

    @dispatch(list)
    def __init__(self, matrix: List[List[float]]) -> None:
        self.__matrix = copy.deepcopy(matrix)
        self.__rows = len(matrix)
        self.__columns = len(matrix[0])

    @dispatch(int, int)
    def __init__(self, rows: int, cols: int) -> None:
        if rows < 1 or cols < 1:
            raise Exception("The number of rows and columns must be greater than 0")

        self.__matrix = [[0 for j in range(cols)] for i in range(rows)]
        self.__rows = rows
        self.__columns = cols

    @dispatch(object)
    def __init__(self, other: Matrix) -> None:
        self.__rows = other.rows
        self.__columns = other.columns
        self.__matrix = copy.deepcopy(other.matrix)

    @property
    def rows(self) -> int:
        return self.__rows

    @property
    def columns(self) -> int:
        return self.__columns

    @property
    def matrix(self) -> List[List[float]]:
        return self.__matrix

    @rows.setter
    def rows(self, value: int) -> None:
        if value <= 0:
            raise Exception("The number of rows in the matrix must be greater than 0")

        if value != self.__rows:
            temp_matrix = self.__matrix
            limit = min(value, self.__rows)

            self.__init__(value, self.__columns)

            for i in range(limit):
                for j in range(self.__columns):
                    self.__matrix[i][j] = temp_matrix[i][j]

    @columns.setter
    def columns(self, value: int) -> None:
        if value <= 0:
            raise Exception("The number of columns in the matrix must be greater than 0")

        if value != self.__columns:
            temp_matrix = self.__matrix
            limit = min(value, self.__columns)

            self.__init__(self.__rows, value)

            for i in range(self.__rows):
                for j in range(limit):
                    self.__matrix[i][j] = temp_matrix[i][j]

    def __add__(self, other: Matrix) -> Matrix:
        result = copy.deepcopy(Matrix(self.matrix))
        result.sum_matrix(other)
        return result

    def __iadd__(self, other: Matrix) -> Matrix:
        self.sum_matrix(other)
        return self

    def __sub__(self, other: Matrix) -> Matrix:
        result = copy.deepcopy(Matrix(self.matrix))
        result.sub_matrix(other)
        return result

    def __isub__(self, other: Matrix) -> Matrix:
        self.sub_matrix(other)
        return self

    def __mul__(self, other: Union[Matrix, float]) -> Matrix:
        result = copy.deepcopy(Matrix(self.matrix))
        if isinstance(other, Matrix):
            result.mul_matrix(other)
        else:
            result.mul_number(other)
        return result

    def __imul__(self, other: Union[Matrix, float]) -> Matrix:
        if isinstance(other, Matrix):
            self.mul_matrix(other)
        else:
            self.mul_number(other)
        return self

    def __str__(self) -> str:
        res: str = ''
        for row in self.matrix:
            res += str(row) + '\n'
        return res

    def __eq__(self, other: Matrix) -> bool:
        return self.eq_matrix(other)

    def sum_matrix(self, second_matrix: Matrix) -> None:
        """ sum of two matrices (a + b) """
        if self.rows != second_matrix.rows or self.columns != second_matrix.columns:
            raise Exception("Matrices can only be added if the dimensions are the same")

        for i in range(self.rows):
            for j in range(self.columns):
                self.matrix[i][j] += second_matrix.matrix[i][j]

    def sub_matrix(self, second_matrix: Matrix) -> None:
        """ subtraction of two matrices (a - b) """
        if self.rows != second_matrix.rows or self.columns != second_matrix.columns:
            raise Exception("Matrices can only be subtracted if the dimensions are the same")

        for i in range(self.rows):
            for j in range(self.columns):
                self.matrix[i][j] -= second_matrix.matrix[i][j]

    def eq_matrix(self, second_matrix: Matrix) -> bool:
        """ check equal matrices (a == b) """
        if self.rows != second_matrix.rows or self.columns != second_matrix.columns:
            return False

        for i in range(self.rows):
            for j in range(self.columns):
                if abs(self.matrix[i][j] - second_matrix.matrix[i][j]) > 1e-6:
                    return False

        return True

    def mul_matrix(self, second_matrix: Matrix) -> None:
        """ multiplication of two matrices (a * b) """
        if self.columns != second_matrix.rows:
            raise Exception("Only (m, k) and (k, n) dimension matrices can be multiplied")

        res_matrix: List[List[float]] = [[0 for _ in range(second_matrix.columns)] for _ in range(self.rows)]

        for i in range(self.rows):
            for j in range(second_matrix.columns):
                for k in range(self.columns):
                    res_matrix[i][j] += self.matrix[i][k] * second_matrix.matrix[k][j]

        self.__init__(res_matrix)

    def mul_number(self, number: float) -> None:
        """ multiplication of matrix and number (a * const) """
        for i in range(self.rows):
            for j in range(self.columns):
                self.matrix[i][j] *= number

    def transpose(self) -> Matrix:
        """ matrix transpose (a ^T) """
        res_matrix: List[List[float]] = [[0 for _ in range(self.rows)] for _ in range(self.columns)]

        for i in range(self.columns):
            for j in range(self.rows):
                res_matrix[i][j] = self.matrix[j][i]

        return Matrix(res_matrix)

    def det(self) -> float:
        """ matrix determinant (|a|) """
        if self.rows != self.columns:
            raise Exception("The determinant can only be calculated for a square matrix")

        return self.__recursive_det(self)

    def calc_complements(self) -> Matrix:
        """ complement matrix """
        if self.rows != self.columns:
            raise Exception("The algebraic complement matrix can only be found for a square matrix")

        res_matrix: List[List[float]] = [[0 for _ in range(self.columns)] for _ in range(self.rows)]

        if self.rows == 1:
            res_matrix[0][0] = 1
        else:
            for i in range(self.rows):
                for j in range(self.columns):
                    matrix_copy = self.__matrix_cut_copy(self, i, j)
                    res_matrix[i][j] = ((-1) ** (i + j)) * self.__recursive_det(matrix_copy)

        return Matrix(res_matrix)

    def inverse_matrix(self) -> Matrix:
        """ inverse matrix (a ^-1) """
        if self.rows != self.columns:
            raise Exception("The inverse matrix can only be found for a square matrix")

        if self.det() == 0:
            raise Exception("The inverse matrix can only be found in the case of a non-zero determinant")

        return (self.calc_complements() * (1 / self.det())).transpose()

    def __recursive_det(self, a: Matrix) -> float:
        """ private method for finding determinant """
        if a.rows == 1:
            return a.matrix[0][0]
        if a.rows == 2:
            return a.matrix[0][0] * a.matrix[1][1] - a.matrix[0][1] * a.matrix[1][0]

        res: float = 0
        for i in range(a.rows):
            copy_matrix = self.__matrix_cut_copy(a, 0, i)
            res += ((-1) ** i) * a.matrix[0][i] * self.__recursive_det(copy_matrix)

        return res

    @staticmethod
    def __matrix_cut_copy(a: Matrix, skip_row: int, skip_column: int) -> Matrix:
        """ private method for finding determinant and complements """
        res_matrix: List[List[float]] = [[0 for _ in range(a.rows - 1)] for _ in range(a.columns - 1)]

        i_new: int = 0
        for i in range(a.rows):
            if i == skip_row:
                continue
            j_new: int = 0
            for j in range(a.columns):
                if j == skip_column:
                    continue
                res_matrix[i_new][j_new] = a.matrix[i][j]
                j_new += 1
            i_new += 1

        return Matrix(res_matrix)

File: Matrices_CPP/src/Matrix.cpp
#include "Matrix.h"

Matrix Matrix::MatrixCutCopy(Matrix A, int row, int column) {
  Matrix result = Matrix(A.rows_ - 1, A.cols_ - 1);

  int i_new = 0;
  int j_new = 0;

  for (int i = 0; i < A.rows_; i++) {
    if (i == row) continue;
    j_new = 0;
    for (int j = 0; j < A.cols_; j++) {
      if (j == column) continue;
      result.matrix_[i_new][j_new++] = A.matrix_[i][j];
    }
    i_new++;
  }
  return result;
}

double Matrix::RecursiveDeterminant(Matrix A) {
  if (A.rows_ == 1) return A.matrix_[0][0];
  if (A.rows_ == 2)
    return A.matrix_[0][0] * A.matrix_[1][1] -
           A.matrix_[0][1] * A.matrix_[1][0];

  double res = 0;
  for (int i = 0; i < A.cols_; i++) {
    Matrix copy = MatrixCutCopy(A, 0, i);
    res += pow(-1, i) * A.matrix_[0][i] * RecursiveDeterminant(copy);
  }
  return res;
}

Matrix::Matrix() {
  this->rows_ = 1;
  this->cols_ = 1;

  this->matrix_ = new double*[1];
  for (int i = 0; i < 1; i++) this->matrix_[i] = new double[1];

  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++) this->matrix_[i][j] = 0;
};

Matrix::~Matrix() { this->Free(); }

Matrix::Matrix(int rows, int cols) {
  if (rows <= 0 || cols <= 0) throw std::logic_error("Incorrect matrix");

  this->rows_ = rows;
  this->cols_ = cols;

  this->matrix_ = new double*[rows];
  for (int i = 0; i < rows; i++) this->matrix_[i] = new double[cols];

  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++) this->matrix_[i][j] = 0;
}

Matrix::Matrix(int rows, int cols, double** matrix) {
  this->rows_ = rows;
  this->cols_ = cols;

  this->matrix_ = new double*[rows];
  for (int i = 0; i < rows; i++) this->matrix_[i] = new double[cols];

  for (int i = 0; i < rows; i++)
    for (int j = 0; j < cols; j++) this->matrix_[i][j] = matrix[i][j];
}

Matrix::Matrix(const Matrix& other) { this->Copy(other); }

Matrix::Matrix(Matrix&& other) {
  this->Copy(other);
  other.Free();
}

double& Matrix::operator()(int row, int col) const {
  if (row >= this->rows_ || col >= this->cols_ || row < 0 || col < 0)
    throw std::logic_error("Incorrect input, index is out of range");

  return matrix_[row][col];
}

Matrix Matrix::operator+(const Matrix& other) const {
  Matrix temp = Matrix(*this);
  temp.SumMatrix(other);
  return temp;
}

Matrix Matrix::operator-(const Matrix& other) const {
  Matrix temp = Matrix(*this);
  temp.SubMatrix(other);
  return temp;
}

Matrix Matrix::operator*(const Matrix& other) const {
  Matrix temp = Matrix(*this);
  temp.MulMatrix(other);
  return temp;
}

Matrix Matrix::operator*(double other) const {
  Matrix temp = Matrix(*this);
  temp.MulNumber(other);
  return temp;
}

Matrix& Matrix::operator+=(const Matrix& other) {
  this->SumMatrix(other);
  return *this;
}

Matrix& Matrix::operator-=(const Matrix& other) {
  this->SubMatrix(other);
  return *this;
}

Matrix& Matrix::operator*=(const Matrix& other) {
  this->MulMatrix(other);
  return *this;
}

Matrix& Matrix::operator*=(double other) {
  this->MulNumber(other);
  return *this;
}

bool Matrix::operator==(const Matrix& other) { return this->EqMatrix(other); }

Matrix& Matrix::operator=(const Matrix& other) {
  *this = Matrix(other);
  return *this;
}

void Matrix::SumMatrix(const Matrix& other) {
  if (this->rows_ != other.GetCols())
    throw std::logic_error(
        "Matrices can only be added if the dimensions are the same");

  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++) this->matrix_[i][j] += other(i, j);
}

void Matrix::SubMatrix(const Matrix& other) {
  if (this->rows_ != other.GetCols())
    throw std::logic_error(
        "Matrices can only be subtracted if the dimensions are the same");

  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++) this->matrix_[i][j] -= other(i, j);
}

void Matrix::MulNumber(const double num) {
  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++) this->matrix_[i][j] *= num;
}

bool Matrix::EqMatrix(const Matrix& other) {
  if (this->rows_ != other.rows_ || this->cols_ != other.cols_) return false;

  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++)
      if (abs(this->matrix_[i][j] - other(i, j)) > 1e-6) return false;

  return true;
}

Matrix Matrix::Transpose() {
  Matrix resultMatrix = Matrix(this->cols_, this->rows_);

  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++)
      resultMatrix(j, i) = this->matrix_[i][j];

  return resultMatrix;
}

void Matrix::MulMatrix(const Matrix& other) {
  if (this->cols_ != other.rows_)
    throw std::logic_error(
        "Only (m, k) and (k, n) dimension matrices can be multiplied");

  Matrix temp = Matrix(this->rows_, other.GetCols());

  for (int i = 0; i < temp.GetRows(); i++)
    for (int j = 0; j < temp.GetCols(); j++)
      for (int k = 0; k < this->cols_; k++)
        (temp)(i, j) += (*this)(i, k) * other(k, j);

  this->Free();
  this->Copy(temp);

  temp.Free();
}

double Matrix::Determinant() {
  if (this->cols_ != this->rows_)
    throw std::logic_error(
        "The determinant can only be calculated for a square matrix");

  double result = RecursiveDeterminant(*this);
  return result;
}

Matrix Matrix::CalcComplements() {
  if (this->rows_ != this->cols_)
    throw std::logic_error(
        "The algebraic complement matrix can only be found for a square "
        "matrix");

  Matrix result = Matrix(this->rows_, this->cols_);

  if (this->rows_ == 1)
    result.matrix_[0][0] = 1;
  else {
    for (int i = 0; i < this->rows_; i++) {
      for (int j = 0; j < this->cols_; j++) {
        Matrix copy = MatrixCutCopy((*this), i, j);
        result(i, j) = pow(-1, i + j) * RecursiveDeterminant(copy);
      }
    }
  }

  return result;
}

Matrix Matrix::InverseMatrix() {
  if (this->rows_ != this->cols_)
    throw std::logic_error(
        "The inverse matrix can only be found for a square matrix");

  double det = this->Determinant();
  if (det == 0)
    throw std::logic_error(
        "The inverse matrix can only be found in the case of a non-zero "
        "determinant");

  Matrix result = this->CalcComplements().Transpose();
  result *= (1 / det);
  return result;
}

void Matrix::Print() {
  for (int i = 0; i < this->rows_; i++) {
    for (int j = 0; j < this->cols_; j++) {
      std::cout << this->matrix_[i][j];
      if (j != this->cols_ - 1) std::cout << ", ";
    }
    std::cout << std::endl;
  }
}

void Matrix::Free() {
  if (this->matrix_ != nullptr) {
    for (int i = 0; i < this->rows_; i++) delete[] this->matrix_[i];

    delete[] this->matrix_;
    this->matrix_ = nullptr;
    this->rows_ = 0;
    this->cols_ = 0;
  }
}

void Matrix::Copy(const Matrix& other) {
  this->rows_ = other.rows_;
  this->cols_ = other.cols_;

  this->matrix_ = new double*[other.rows_];
  for (int i = 0; i < other.rows_; i++)
    this->matrix_[i] = new double[other.cols_];

  for (int i = 0; i < this->rows_; i++)
    for (int j = 0; j < this->cols_; j++) this->matrix_[i][j] = other(i, j);
}

int Matrix::GetRows() const { return this->rows_; }

int Matrix::GetCols() const { return this->cols_; }

void Matrix::SetRows(int newRows) {
  if (newRows < 0)
    throw std::logic_error(
        "The number of rows in the matrix must be greater than 0");

  if (newRows != this->rows_) {
    Matrix tmp = Matrix(newRows, this->cols_);
    int min = std::min(this->rows_, newRows);

    for (int i = 0; i < min; i++)
      for (int j = 0; j < this->cols_; j++) tmp(i, j) = (*this)(i, j);

    this->Free();
    this->Copy(tmp);
  }
}

void Matrix::SetCols(int newCols) {
  if (newCols < 0)
    throw std::logic_error(
        "The number of columns in the matrix must be greater than 0");

  if (newCols != this->cols_) {
    Matrix tmp = Matrix(this->rows_, newCols);
    int min = std::min(this->cols_, newCols);

    for (int i = 0; i < this->rows_; ++i)
      for (int j = 0; j < min; ++j) tmp(i, j) = (*this)(i, j);

    this->Free();
    this->Copy(tmp);
  }
}
